int8    state
float64 rotation
